### PR TITLE
Improvement in electrostic field solve when using EB & a small refactoring of the Electrostatic solver

### PR DESCRIPTION
**Pull request by roelof-groenewald**
This PR reorganizes some parts of `ElectrostaticSolver.cpp` with two goals:
1. Use the AMReX `mlmg.getGradSolution` functionality to accurately calculate the electric field in a system with an embedded boundary. (Note that this caused the reference values for the Efield in test `ElectrostaticSphereEB` to change).
2. Allow an outside (Python) Poisson solver to be installed - the callback `installpoissonsolver` has been added for this.

Some changes here overlap with PR #1811.

### CODE REVIEW

**Comment by peterscherpelz** [comment]
@RemiLehe FYI, @roelof-groenewald is out today and tomorrow, but the rest of the team wanted to point out the title should probably be changed away from refactoring. While that's part of this PR, the gradient solution does actually change physics for the better - we hit significant problems with the E-field calculation using just finite difference in an EB case - so this is really a bugfix or improvement.

File: Source/WarpX.cpp
Comment (dpgrote) [inline Source/WarpX.cpp]: As the code is now, this should be `do_electrostatic == ElectrostaticSolverAlgo::LabFrame` since the other option still uses its own local copy of rho.

File: Source/FieldSolver/ElectrostaticSolver.cpp
Comment (dpgrote) [inline Source/FieldSolver/ElectrostaticSolver.cpp]: Note that PR #1811 adds a call to SynchRho after calling `species.DepositCharge` (and effectively sets local=false). As noted there, this allows the option of applying filtering to the charge density. Note that that PR removes the `local` option since it is always false.

Comment (PhilMiller) [reply Source/FieldSolver/ElectrostaticSolver.cpp]: That PR has been idle since April. Is it expected to move forward?

Comment (PhilMiller) [reply Source/FieldSolver/ElectrostaticSolver.cpp]: @RemiLehe since that's your PR

**Comment by PhilMiller** [comment]
The test failures are differences in the Electrostatic Sphere. I assume that probably needs to be updated to take the new results as the reference value

File: Source/FieldSolver/ElectrostaticSolver.cpp
Comment (WeiqunZhang) [inline Source/FieldSolver/ElectrostaticSolver.cpp]: Do we need to multiply `Ey` by `-1` in 2D?

Comment (roelof-groenewald) [reply Source/FieldSolver/ElectrostaticSolver.cpp]: Nope, that is definitely just a performance hit. I've now fixed it. Thanks for catching it!

File: Source/FieldSolver/ElectrostaticSolver.cpp
Comment (PhilMiller) [inline Source/FieldSolver/ElectrostaticSolver.cpp]: This `+` is unnecessary and I think wrong. The C++ preprocessor will automatically concatenate adjacent string literals. Otherwise, you're doing pointer addition of the two resulting `const char*` values, which would crash or print gibberish while attempting to report the assertion failure. Try it and see?

Comment (roelof-groenewald) [reply Source/FieldSolver/ElectrostaticSolver.cpp]: Thanks for the comment. I now tested it and it works fine with the `+` but it actually didn't achieve what I was hoping it would so I reverted the change. I've noticed that the error messages in  `AMREX_ALWAYS_ASSERT_WITH_MESSAGE` don't print as a single string when they are broken up over multiple lines such as the above case. For example triggering the above gives output:

```0::Assertion `false' failed, file "./Source/FieldSolver/ElectrostaticSolver.cpp", line 413, Msg: "Field boundary conditions have to be either periodic or PEC " + "when using the electrostatic solver" !!!```

and similarly without the `+` it gives:

```0::Assertion `false' failed, file "./Source/FieldSolver/ElectrostaticSolver.cpp", line 413, Msg: "Field boundary conditions have to be either periodic or PEC " "when using the electrostatic solver" !!!```

**Review by WeiqunZhang — APPROVED** [review]

File: Source/FieldSolver/ElectrostaticSolver.cpp
Comment (RemiLehe) [reply Source/FieldSolver/ElectrostaticSolver.cpp]: I think that PR #1811 might still move forward. 
Also, if I understood correctly, the changes related to `DepositChargeDensity` and `ChargeDensityGridProcessing` in the current PR are independent from the other changes and are purely a refactoring of the charge deposition. Is that correct?

If so, I would suggest to revert the changes associated with `ChargeDensityGridProcessing` and `DepositChargeDensity` from the current PR, so as to avoid a conflict with #1811

**Review by RemiLehe — CHANGES_REQUESTED** [review]
Thanks a lot for this PR, and sorry for the late review!

I have a minor request regarding the changes to the charge deposition (see inline discussion).

File: Source/FieldSolver/ElectrostaticSolver.cpp
Comment (roelof-groenewald) [reply Source/FieldSolver/ElectrostaticSolver.cpp]: Thanks for reviewing @RemiLehe. The refactoring to create `DepositChargeDensity` and `ChargeDensityGridProcessing` is just to break out the steps involved in the field solve somewhat more clearly and should hopefully not interfere too much with the other PR. The part that definitely does create a conflict is the switch to using `rho_fp` for the lab frame solution. We need this though for the external solver that we use to work (the reason for the `installpoissonsolver` part) since the `rho_fp` multifabs can be accessed from Python.
So I see three possible routes to clear things up:
1. Have this PR wait for #1811 to be merged, then update it with those changes afterwards
2. We split this out into 2 PR's: one for the `mlmg.getGradSolution` stuff now, and one for the `installpoissonsolver` and refactor stuff after 1811 is merged
3. Merge this one as is and update 1811 with these changes - I know this is the option that creates the least work for us and the most work for you, but I think it might be the simplest since it would basically just remove the `ChargeDensityGridProcessing` function and replace calls to it with a call to `SyncRho();` as well as remove the `local` flag from `DepositCharge`. If you'd like I'm happy to make those changes in a follow-up PR that will close 1811 as well.

Please let me know which route you think is best or if there is a different approach that would be even better.

Also just for your awareness the refactoring of the charge deposition is useful for us since it allows us to get the charge density for a specific species deposited on the grid and saved in `rho_fp` at any point during the simulation and then access that charge density from Python. We sometimes use this as a diagnostic.

File: Regression/Checksum/benchmarks_json/ElectrostaticSphereEB.json
Comment (roelof-groenewald) [inline Regression/Checksum/benchmarks_json/ElectrostaticSphereEB.json:5]: Note that these values changed due to the use of `mlmg.getGradSolution` to solve for the E-field when embedded boundaries are present, which corrects the less accurate finite difference method used before (that didn't incorporate cut-cells).

File: Source/FieldSolver/ElectrostaticSolver.cpp
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp]: My impression is that this function is not currently used in the code below, is that correct? If so, I would suggest to remove it.
```suggestion
```

File: Source/FieldSolver/ElectrostaticSolver.cpp
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp]: Could you explain  why this part is now done outside of `AddSpaceChargeFieldLabFrame`?
If it is actually equivalent to do it inside `AddSpaceChargeFieldLabFrame`, then I would prefer doing so.

File: Source/FieldSolver/ElectrostaticSolver.cpp
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp:484]: Could we actually put this code inside of the function `computeE` ? I think it would be slightly more consistent.

File: Source/WarpX.H
Comment (RemiLehe) [inline Source/WarpX.H]: It seems that this function is actually not used in the code. Could we remove it?

**Review by RemiLehe — CHANGES_REQUESTED** [review]
Thanks for updating this PR! I have a few additional comments here.

File: Source/FieldSolver/ElectrostaticSolver.cpp
Comment (roelof-groenewald) [reply Source/FieldSolver/ElectrostaticSolver.cpp]: It is true that it is not currently used in `ElectrostaticSolver.cpp` but we do use it in order to get species specific charge density during diagnostics which is why it is there. I'm not sure where a better place would be to put it but happy to take suggestions.

File: Source/FieldSolver/ElectrostaticSolver.cpp
Comment (roelof-groenewald) [reply Source/FieldSolver/ElectrostaticSolver.cpp:484]: I agree that would be nicer, but it requires the `mlmg` object and I wasn't able to think of a simple and clean way to pass that object along to `computeE`. If you have a suggestion of how to do that I'd be happy to follow it.

File: Source/FieldSolver/ElectrostaticSolver.cpp
Comment (roelof-groenewald) [reply Source/FieldSolver/ElectrostaticSolver.cpp]: Actually nevermind, I found a workaround. Will update shortly.

File: Source/FieldSolver/ElectrostaticSolver.cpp
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp:141]: Could this part be added back into the `AddSpaceChargeFieldLabFrame` (instead of being done before calling it)?

File: Source/FieldSolver/ElectrostaticSolver.cpp
Comment (RemiLehe) [reply Source/FieldSolver/ElectrostaticSolver.cpp:484]: OK, agreed. We can leave it as is for now.

**Review by RemiLehe — CHANGES_REQUESTED** [review]
Thanks for updating the PR again! I have one additional minor request (see inline comments)

**Comment by PhilMiller** [comment]
The macOS CI failure seems to have been a system problem, in which it couldn't reach Github to download. Either re-run that, or merge in spite of it.

**Review by RemiLehe — APPROVED** [review]

**Comment by RemiLehe** [comment]
Thank you so much for this PR, and sorry for stalling it for so long!

**Pull request merged**
